feat: let user select pr to checkout

nilvng opened the pull request:
fixes #2329

## Acceptance criteria
If no argument specified for `pr checkout` and we're attached to a TTY, 
- fetch open PRs and 
- offer them in a `prompter.Select`

## Functionality:
- When the user attempts a PR checkout without specifying a PR number, and the user is attached to a TTY (text terminal), this prompt will automatically activate. The prompt will display a list of **10 most recent PRs** (tentative) from the base repository
Users can then select the desired PR from the list using the arrow keys and confirm selection with Enter.
- Two new unit tests are included
## Iteration history
TTY

https://github.com/user-attachments/assets/499b5dfb-3103-42b8-876a-3a2d4d7173c8

Non TTY
<img width="840" alt="Screenshot 2024-12-06 at 7 44 31 pm" src="https://github.com/user-attachments/assets/43de7015-19c8-4356-9251-32ae81c579f8">

I'd like to thank @BagToad  for their constructive feedback throughout the development of this feature.

<details>
<summary>Second iteration</summary>
![2024-11-18 00 03 17](https://github.com/user-attachments/assets/c784fb05-fafd-4b3e-87f9-0948bdb61092)
</details>

<details>
<summary>First iteration</summary>
<img width="505" alt="Screenshot 2024-11-03 at 11 10 58 AM" src="https://github.com/user-attachments/assets/358d3d61-b657-4ede-94ca-33e20d18f2b4">
</details>

nilvng committed:
feat: let user select pr to checkout

nilvng commented on pkg/cmd/pr/checkout/checkout.go:
```diff
@@ -258,3 +289,40 @@ func executeCmds(client *git.Client, cmdQueue [][]string) error {
+		"isCrossRepository",
+		"isDraft",
+		"createdAt",
+	}}, 30)
```
From personal experience, our team is quite small, and we tidy up our PR queue quickly so I rarely need more than the first 30 PRs. 

But let me know how we want to design this feature, to cover your use cases also :)

BagToad replied:
I think 10-20 might be a reasonable number to start with, which I see you've gone with now 👍 

For what it's worth, `gh run view` only shows `10` runs as well:

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/run/view/view.go#L237-L237

Recognizing that workflow runs are different than pull requests, but I bring it up as a reference point example to help us understand how many items users are typically interested in seeing in interactive prompts 😁

nilvng committed:
Merge branch 'trunk' into nil/fix-2329

nilvng committed:
issue #2329: fix the linting issue

nilvng committed:
issue #2329: fix the unit tests issue - ensure its backward compatible

nilvng committed:
issue #2329: add happy path unit test

nilvng committed:
issue #2329: improve UI/UX

Improve the interactive PR selection UI by
- prefix the PR number with hashcode #
- perserve the text formatting (bold) upon an option is hovered
- add the PR head label

Technical changes:
- Replace \033[0m with \033[39m for maintaining text formatting

BagToad committed:
Merge branch 'trunk' into nil/fix-2329

BagToad commented on pkg/cmd/pr/shared/display.go:
```diff
@@ -17,6 +17,15 @@ func StateTitleWithColor(cs *iostreams.ColorScheme, pr api.PullRequest) string {
+// based on its state (open, closed, merged, or draft).
+// It prefixes the number with a hash symbol (#) to indicate it's a pull request.
+func PRNumberWithColor(cs *iostreams.ColorScheme, pr api.PullRequest) string {
+	prStateColorFunc := cs.ColorFromRGB(ColorHexCodeForPRState(pr))
```
🤔 I have some accessibility concerns with this and `ColorHexCodeForPRState`. I believe this overrides the user's terminal color scheme with these raw hex codes, which can cause some unwanted behavior. For example, here's what this existing implementation looks like in my terminal:

![image](https://github.com/user-attachments/assets/9cdfe440-a32b-4a31-a504-535b907625e2)

I think the way we'd like this instead is to use the user's color terminal color scheme, which we can leverage the existing `ColorForPRState` function for:

```suggestion
	prStateColorFunc := cs.ColorFromString(ColorForPRState(pr))
```

Here's what this change looks like in my terminal ❤️ :

![image](https://github.com/user-attachments/assets/42f1b582-5480-4784-9523-908f8b2c76e0)

This also has the added benefit of [aligning exactly with the colors used by `gh pr list`](https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/pr/list/list.go#L221):

![image](https://github.com/user-attachments/assets/cadd4e88-96c4-4b85-be41-05b28c940a92)

Here is our [Primer design resource](https://primer.style/native/cli/foundations#color) for terminal colors as well as a reference ✨ 

With all this in mind, I propose we call `ColorForPRState` instead and remove `ColorHexCodeForPRState` 😁

nilvng replied:
Hey thanks for testing this out! I hadn't considered the issue of hex codes overwriting the system theme – appreciate the detail on that!

The reason for my changes was to avoid using the existing `ColorForPRState` function. It relies on the `mguz/ansi` library internally, which has a potential issue. Their ColorFunc uses the ANSI reset code` \033[0m` which resets all text styles, not just the color we set. Additionally, their suggestion to manually build the ANSI escape sequence wouldn't work for the same reason – it still resets styles beyond just color.

```go
fn = func(s string) string {
	// return fmt.Sprint("\033[2m", s, "\033[0m") // expect: a s string color in green
	 return fmt.Sprint(ansi.Green, s, "\033[0m") // actual: style string color in green and normal text weight
}
```
(Here are some reference links for the library behavior if you're interested: [link 1](https://github.com/mgutz/ansi/blob/d51e80ef957dba7f19388ca64afefbd5a096af30/ansi.go#L245C1-L247C25) and [link 2](https://github.com/mgutz/ansi/blob/d51e80ef957dba7f19388ca64afefbd5a096af30/ansi.go#L169C1-L170C1))

In our case, the `Selector.Prompt()` use a bold style for the current line, and unfortunately, the `ColorForPRState` behavior resets that bold style.


| PR attempt to fix the text style | Revert |
|---|---|
|<img width="1005" alt="Screenshot 2024-11-30 at 8 55 35 pm" src="https://github.com/user-attachments/assets/e2385c8a-37a2-41ea-b7b7-7c2f34a0a615">| <img width="706" alt="Screenshot 2024-11-30 at 9 00 34 pm" src="https://github.com/user-attachments/assets/2f7f2ffd-6c7a-4a23-8482-9833c53f6c12"> |


I was wondering if there is something we can do from our end to keep the bold style, because personally I find it easier to follow when the PR title, and branch name is very long - but tbh I don't have strong opinion either way so I happy to revert the changes to `\033[0m` and also use the `ColorForPRState()` here instead :) For the issue with the library, I could potentially raise it in their repo, and propose a fix there  :)

let me what you think about it!

BagToad replied:
I see! Excellent and interesting investigation 🤔 

I'd also like to keep the bolding if we can 💯  I think that's a good feature. I think there is likely something we can do here to keep the bolding. I'll think about it more 💭  

No matter what, I think we will need to align with the colors `gh pr list` is using and ensure `gh` remains accessible by respecting terminal color preferences.

BagToad replied:
Thinking about this more, I couldn't find any [existing code that leverages colors in a `prompter.Select` invocation](https://github.com/search?q=repo%3Acli%2Fcli+prompter.Select%28&type=code) 🤔 

I think an option here is to _remove all color from the prompt_ except for the currently selected item, aligning with how other prompts seem to work.

Right now, even some prompts where color might make sense do not currently have color. Take `gh run view` for example:

![image](https://github.com/user-attachments/assets/93534304-600d-48d9-9534-0541077a1c76)

I could see an argument that we may want to color the `X` and `✓` icons. 

So, _@nilvng what do you think about removing color altogether for now, and maybe opening a follow-up issue/PR to resolve this color issue, improving the availability of color in prompting generally?_


Sidenote - I'm also reflecting on Primer design guidelines, and what that might mean for color in prompts 🤔 :

> Only use color to [enhance meaning](https://primer.style/components/icon#usage), not to communicate meaning.

nilvng replied:
Right, That makes sense! It seems we've consistently avoided color in other prompts.

For now, I think its all good for us to remove color from this prompt entirely. While I'd like to use gray and green to categorize PRs, the new UI doesn't align with our previous prompt design. Plus, I could also see that using color here have cross the line of enhancing to communicating  e.g. PR number that come with draft and in review status

BagToad commented on pkg/cmd/pr/checkout/checkout.go:
```diff
@@ -268,3 +312,44 @@ func executeCmds(client *git.Client, credentialPattern git.CredentialPattern, cm
+		)))
+	}
+
+	selected, err := prompter.Select("Check out a specific PR?", "", candidates)
```
This suggestion is to align this prompt text with some other "select a `<something>`" commands:

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/workflow/shared/shared.go#L201

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/gist/shared/shared.go#L241

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/pr/shared/templates.go#L202

I believe we also want to avoid abbreviating terms like `pull request` when they are part of a prompt or help message to make it [feel like GitHub](https://primer.style/native/cli/getting-started#make-it-feel-like-github).

```suggestion
	selected, err := prompter.Select("Select a pull request", "", candidates)
```

nilvng replied:
Updated! thanks 🫡

BagToad commented on pkg/iostreams/color.go:
```diff
@@ -252,5 +252,5 @@ func (c *ColorScheme) HexToRGB(hex string, x string) string {
 	g, _ := strconv.ParseInt(hex[2:4], 16, 64)
 	b, _ := strconv.ParseInt(hex[4:6], 16, 64)
-	return fmt.Sprintf("\033[38;2;%d;%d;%dm%s\033[0m", r, g, b, x)
+	return fmt.Sprintf("\033[38;2;%d;%d;%dm%s\033[39m", r, g, b, x)
```
Could you talk more about the reason for this change? 👀

nilvng replied:
I was thinking because this was supposed to set the foreground color, the reset code should only reset the foreground color `\033[39m`, but not the hard reset `\033[0m` for all styles - which might stop other style like bold, italic apply to the rest of the string

BagToad commented on pkg/cmd/pr/list/http.go:
```diff
@@ -13,7 +13,7 @@ func shouldUseSearch(filters prShared.FilterOptions) bool {
 }
 
-func listPullRequests(httpClient *http.Client, repo ghrepo.Interface, filters prShared.FilterOptions, limit int) (*api.PullRequestAndTotalCount, error) {
+func ListPullRequests(httpClient *http.Client, repo ghrepo.Interface, filters prShared.FilterOptions, limit int) (*api.PullRequestAndTotalCount, error) {
```
🤔 Since this is now exported and used by multiple commands, I wonder if it should be extracted into a file like https://github.com/cli/cli/blob/trunk/api/queries_pr.go

We did something similar in #9721 , [migrating `listLicenseTemplates` to `api.RepoLicenses`](https://github.com/cli/cli/pull/9721/files#diff-1383afb03d1730b7f23e2da529e535aa1fc1c5ac500fc802899dafa66fcbb61eL247-R247)

What do you think?

nilvng replied:
Yeah, I totally agree we should refactor this function. Ito make it evident that it's a shared component used by multiple commands, not solely `pr list`

I was a bit hesitant at first because I'm still getting used to the codebase, so I don't want to alter too much before getting guidance from the team ☺️ 

 thanks for the inspo!  I will probably open a new PR branched off this PR for the refactoring - to keep this PR focus on the discussion around the feature itself - let me what you think!

BagToad replied:
Sure, a new PR merging into this branch makes sense to discuss that refactor 👍

BagToad reviewed (changes requested):
👋 Hey @nilvng ✨ 

Thank you for taking on this old feature request. I am personally looking forward to this feature 😁 ❤️ 

I added a few comments on the code for things that I think we need to address. I think the team and I may have more comments later, but this should get us moving 😃  

I've also noted a few things generally below...

----

I think we need to address the inverse of the acceptance criteria; prompting should _only_ happen when we are attached to a TTY. 

Here is the current behavior when run non-interactively:

![image](https://github.com/user-attachments/assets/81283095-eadf-43de-b7b9-cc0ed62de454)

We must not prompt in this case. Here is an example from `gh workflow run` showing the desired behavior:

![image](https://github.com/user-attachments/assets/05058c8c-32af-42d5-9f88-470b377f9aa2)

Here is the implementation that we can pull inspiration from:

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/workflow/run/run.go#L98-L99

-----

There are also some unit test failures that need to be investigated, though I suspect it may be due to some conflict with the `trunk` branch changes I merged back into your branch 🤔

nilvng committed:
issue #2329: fix missing stub

nilvng committed:
issue #2329: return error when not running interactively

nilvng reviewed (commented):
Thanks a lot @BagToad for the feedback! 
I'm a heavy CLI user at work and have been eagerly awaiting this feature. Ngl, the PR initially put up with my pure enthusiasm for the feature was quite sparse. I'm keen to learn from your expertise and guidance on my approach.

BagToad commented on pkg/cmd/pr/checkout/checkout.go:
```diff
@@ -71,15 +87,47 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
+		}
+
+	default:
+		if !opts.IO.CanPrompt() {
+			return cmdutil.FlagErrorf("must provide a pull request number (or URL or branch) when not running interactively")
+		}
```
I think we should move this logic into the `RunE` section of the Cobra setup to align with some of the other commands. We could also reword the error message to align with the other commands:

```
pull request number, URL, or branch required when not running interactively
```

Here's some examples:

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/run/view/view.go#L155-L163

https://github.com/cli/cli/blob/9c550992ea4f207fa293cf0de00b633b47cff4c9/pkg/cmd/workflow/run/run.go#L96-L102

BagToad replied:
Note: this is also resolved through my suggested patch in https://github.com/cli/cli/pull/9868#issuecomment-2513244933

BagToad commented:
@nilvng Thanks for your continued work on this and insights into your thought process ✨ 

I've been reviewing this a bit deeper and have a few more things I'd like to address, and I also have a proposal for a refactor.

### Interactive checkouts can behave unexpectedly for fork PRs

_TLDR: we must include both `headRepository` and `maintainerCanModify` in the requested fields even though we don't display those fields to the user when they select the PR. We need to request at least the same fields as the non-interactive flow to later do the necessary Git operations_ 

While testing this, I've noticed that some PRs cause my terminal prompt to behave unexpectedly, and clue me into that the interactive flow was doing different `git` operations than the non-interactive flow:

![image](https://github.com/user-attachments/assets/1c902f9a-3e37-4941-a3a5-6d0505170d9c)

This is what I'd typically expect instead:

![image](https://github.com/user-attachments/assets/c768b3fd-bcf1-4c0b-bec8-adb2f3a3e8c7)

I discovered this was because of this logic which occurs later in the process:

https://github.com/cli/cli/blob/d92e529629f2ccc0c87c4a49e0b2f199602bfb3c/pkg/cmd/pr/checkout/checkout.go#L271-L275

Since the interactive flow was not requesting these fields ( `headRepository` and `maintainerCanModify`), they were false or `nil`. This causes the the `.merge` config to be set to a reference to a pull request's head commit instead of a reference to a branch. 

**Interactive flow**

https://github.com/cli/cli/blob/d92e529629f2ccc0c87c4a49e0b2f199602bfb3c/pkg/cmd/pr/checkout/checkout.go#L321-L330

**non-interactive flow**

https://github.com/cli/cli/blob/d92e529629f2ccc0c87c4a49e0b2f199602bfb3c/pkg/cmd/pr/checkout/checkout.go#L98-L108

## Refactoring interactive flow

I'd like to propose a refactor to the logic around deciding whether to prompt. I'd like to:

- Align more closely with other command implementations by moving away from the `switch` statement for this use-case.
- Improve readability by moving more logic into `selectPR`, which I've renamed to `ResolvePR` given it's new behavior.

Note: this proposal includes solutions to the above problem regarding the `headRepository` and `maintainerCanModify` fields and solutions to [my other comment](https://github.com/cli/cli/pull/9868#discussion_r1866407659) about moving the prompting check into `RunE`.

I'd like to push a commit to this PR with my proposal below, if you agree with the essence of the refactor, then we can refactor further if needed ❤️ 

Please feel free to `git apply` this diff below to your local branch and let me know what you think 😁 🙌  

Edit: here's [an example from `gh workflow enable`](https://github.com/cli/cli/blob/5a74934d192ddce979ca64377be6e10af8680467/pkg/cmd/workflow/enable/enable.go#L77) that is similar to the refactor I am going for. In that case `ResolveWorkflow` handles getting a workflow from the user, and it will prompt the user if it needs to - I propose a similar idea with `resolvePR` that handles getting a PR from the user, enclosing the decision to prompt or not within that function.  

<details><summary>Show diff</summary>
<p>

```diff
diff --git a/pkg/cmd/pr/checkout/checkout.go b/pkg/cmd/pr/checkout/checkout.go
index e78eb026..d850f70c 100644
--- a/pkg/cmd/pr/checkout/checkout.go
+++ b/pkg/cmd/pr/checkout/checkout.go
@@ -69,11 +69,14 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
 
 			if len(args) > 0 {
 				opts.SelectorArg = args[0]
+			} else if !opts.IO.CanPrompt() {
+				return cmdutil.FlagErrorf("pull request number, URL, or branch required when not running interactively")
 			}
 
 			if runF != nil {
 				return runF(opts)
 			}
+
 			return checkoutRun(opts)
 		},
 	}
@@ -87,47 +90,21 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
 }
 
 func checkoutRun(opts *CheckoutOptions) error {
-	var (
-		baseRepo ghrepo.Interface
-		pr       *api.PullRequest
-		err      error
-	)
-
-	switch {
-	case opts.SelectorArg != "":
-		pr, baseRepo, err = opts.Finder.Find(shared.FindOptions{
-			Selector: opts.SelectorArg,
-			Fields: []string{
-				"number",
-				"headRefName",
-				"headRepository",
-				"headRepositoryOwner",
-				"isCrossRepository",
-				"maintainerCanModify",
-			},
-		})
-		if err != nil {
-			return err
-		}
-
-	default:
-		if !opts.IO.CanPrompt() {
-			return cmdutil.FlagErrorf("must provide a pull request number (or URL or branch) when not running interactively")
-		}
-
-		httpClient, err := opts.HttpClient()
-		if err != nil {
-			return err
-		}
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
 
-		if baseRepo, err = opts.BaseRepo(); err != nil {
-			return err
-		}
+	baseRepo, err := opts.BaseRepo()
+	if err != nil {
+		return err
+	}
 
-		if pr, err = selectPR(httpClient, baseRepo, opts.Prompter, opts.IO.ColorScheme()); err != nil {
-			return err
-		}
+	pr, err := resolvePR(httpClient, baseRepo, opts.Prompter, opts.SelectorArg, opts.Finder, opts.IO.ColorScheme())
+	if err != nil {
+		return err
 	}
+
 	cfg, err := opts.Config()
 	if err != nil {
 		return err
@@ -317,22 +294,46 @@ func executeCmds(client *git.Client, credentialPattern git.CredentialPattern, cm
 	return nil
 }
 
-func selectPR(httpClient *http.Client, baseRepo ghrepo.Interface, prompter shared.Prompter, cs *iostreams.ColorScheme) (*api.PullRequest, error) {
+func resolvePR(httpClient *http.Client, baseRepo ghrepo.Interface, prompter shared.Prompter, pullRequestSelector string, pullRequestFinder shared.PRFinder, cs *iostreams.ColorScheme) (*api.PullRequest, error) {
+	// When non-interactive
+	if pullRequestSelector != "" {
+		pr, _, err := pullRequestFinder.Find(shared.FindOptions{
+			Selector: pullRequestSelector,
+			Fields: []string{
+				"number",
+				"headRefName",
+				"headRepository",
+				"headRepositoryOwner",
+				"isCrossRepository",
+				"maintainerCanModify",
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		return pr, nil
+	}
+
+	// When interactive
 	listResult, err := list.ListPullRequests(httpClient, baseRepo, shared.FilterOptions{Entity: "pr", State: "open", Fields: []string{
 		"number",
 		"title",
 		"state",
 		"url",
 		"headRefName",
+		"headRepository",
 		"headRepositoryOwner",
 		"isCrossRepository",
 		"isDraft",
 		"createdAt",
+		"maintainerCanModify",
 	}}, 10)
 	if err != nil {
 		return nil, err
 	}
+
 	pr, err := promptForPR(prompter, cs, *listResult)
+
 	return pr, err
 }
 

```
</p>
</details>

nilvng committed:
fix: update the copy

nilvng committed:
issue #2329: move Interactive check to RunE

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

nilvng commented:
Hey @BagToad  How are you doin!
I'm just wanting to give you a quick update:
1. On the missing `MaintainerCanModify` and `HeadRepository` - very appreciate the catch! 
I was hoping that the unit tests would get my back, as it did before, but  seems like there wasn't one in place to catch this particular issue 😓   I expect that the next refactor will enable us to assert the Pull request Fragment similar to Finder() [[ref](https://github.com/nilvng/cli/blob/9c82156e8626486791f9f22165f3127498c43752/pkg/cmd/pr/checkout/checkout_test.go#L462-L463)]. Would you have any suggestions on how to implement effective unit tests for this scenario?

2. Regarding your proposal of `ResolvePR()` - it makes totally sense to me! The new function maintain the code symmetry which is a huge improvement and def would have prevented the bug!  Later after applying the proposed changes the unit tests started failing - I went on to another rabbit holes of unit tests... noticed that  `opts.BaseRepo` was not ever mocked in `checkout_test`, but AFAIK, it supposed to return the same value. Thus, I made following up changes  - keen for your thoughts! 🙌 

<details>
<summary>Show diff</summary>

```diff

diff --git a/pkg/cmd/pr/checkout/checkout.go b/pkg/cmd/pr/checkout/checkout.go
index 9516e580..b9ded029 100644
--- a/pkg/cmd/pr/checkout/checkout.go
+++ b/pkg/cmd/pr/checkout/checkout.go
@@ -92,47 +92,21 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
 }
 
 func checkoutRun(opts *CheckoutOptions) error {
-	var (
-		baseRepo ghrepo.Interface
-		pr       *api.PullRequest
-		err      error
-	)
-
-	switch {
-	case opts.SelectorArg != "":
-		pr, baseRepo, err = opts.Finder.Find(shared.FindOptions{
-			Selector: opts.SelectorArg,
-			Fields: []string{
-				"number",
-				"headRefName",
-				"headRepository",
-				"headRepositoryOwner",
-				"isCrossRepository",
-				"maintainerCanModify",
-			},
-		})
-		if err != nil {
-			return err
-		}
-
-	default:
-		if !opts.Interactive {
-			return cmdutil.FlagErrorf("must provide a pull request number (or URL or branch) when not running interactively")
-		}
-
-		httpClient, err := opts.HttpClient()
-		if err != nil {
-			return err
-		}
+	baseRepo, err := opts.BaseRepo()
+	if err != nil {
+		return err
+	}
 
-		if baseRepo, err = opts.BaseRepo(); err != nil {
-			return err
-		}
+	client, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
 
-		if pr, err = selectPR(httpClient, baseRepo, opts.Prompter, opts.IO.ColorScheme()); err != nil {
-			return err
-		}
+	pr, err := resolvePR(client, baseRepo, opts.Prompter, opts.SelectorArg, opts.Interactive, opts.Finder, opts.IO.ColorScheme())
+	if err != nil {
+		return err
 	}
+
 	cfg, err := opts.Config()
 	if err != nil {
 		return err
@@ -322,22 +296,49 @@ func executeCmds(client *git.Client, credentialPattern git.CredentialPattern, cm
 	return nil
 }
 
-func selectPR(httpClient *http.Client, baseRepo ghrepo.Interface, prompter shared.Prompter, cs *iostreams.ColorScheme) (*api.PullRequest, error) {
+func resolvePR(httpClient *http.Client, baseRepo ghrepo.Interface, prompter shared.Prompter, pullRequestSelector string, isInteractive bool, pullRequestFinder shared.PRFinder, cs *iostreams.ColorScheme) (*api.PullRequest, error) {
+	// When non-interactive
+	if pullRequestSelector != "" {
+		pr, _, err := pullRequestFinder.Find(shared.FindOptions{
+			Selector: pullRequestSelector,
+			Fields: []string{
+				"number",
+				"headRefName",
+				"headRepository",
+				"headRepositoryOwner",
+				"isCrossRepository",
+				"maintainerCanModify",
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		return pr, nil
+	}
+	if !isInteractive {
+		return nil, cmdutil.FlagErrorf("pull request number, URL, or branch required when not running interactively")
+	}
+	// When interactive
 	listResult, err := list.ListPullRequests(httpClient, baseRepo, shared.FilterOptions{Entity: "pr", State: "open", Fields: []string{
 		"number",
 		"title",
 		"state",
 		"url",
+		"isDraft",
+		"createdAt",
+
 		"headRefName",
+		"headRepository",
 		"headRepositoryOwner",
 		"isCrossRepository",
-		"isDraft",
-		"createdAt",
+		"maintainerCanModify",
 	}}, 10)
 	if err != nil {
 		return nil, err
 	}
+
 	pr, err := promptForPR(prompter, cs, *listResult)
+
 	return pr, err
 }
 
diff --git a/pkg/cmd/pr/checkout/checkout_test.go b/pkg/cmd/pr/checkout/checkout_test.go
index 49ac0bb2..5c2109f2 100644
--- a/pkg/cmd/pr/checkout/checkout_test.go
+++ b/pkg/cmd/pr/checkout/checkout_test.go
@@ -85,6 +85,10 @@ func Test_checkoutRun(t *testing.T) {
 					finder := shared.NewMockFinder("123", pr, baseRepo)
 					return finder
 				}(),
+				BaseRepo: func() (ghrepo.Interface, error) {
+					baseRepo, _ := stubPR("OWNER/REPO:master", "OWNER/REPO:feature")
+					return baseRepo, nil
+				},
 				Config: func() (gh.Config, error) {
 					return config.NewBlankConfig(), nil
 				},
@@ -113,6 +117,10 @@ func Test_checkoutRun(t *testing.T) {
 					finder := shared.NewMockFinder("123", pr, baseRepo)
 					return finder
 				}(),
+				BaseRepo: func() (ghrepo.Interface, error) {
+					baseRepo, _ := stubPR("OWNER/REPO:master", "OWNER/REPO:feature")
+					return baseRepo, nil
+				},
 				Config: func() (gh.Config, error) {
 					return config.NewBlankConfig(), nil
 				},
@@ -143,6 +151,10 @@ func Test_checkoutRun(t *testing.T) {
 					finder := shared.NewMockFinder("123", pr, baseRepo)
 					return finder
 				}(),
+				BaseRepo: func() (ghrepo.Interface, error) {
+					baseRepo, _ := stubPR("OWNER/REPO:master", "OWNER/REPO:feature")
+					return baseRepo, nil
+				},
 				Config: func() (gh.Config, error) {
 					return config.NewBlankConfig(), nil
 				},
@@ -171,6 +183,10 @@ func Test_checkoutRun(t *testing.T) {
 					finder := shared.NewMockFinder("123", pr, baseRepo)
 					return finder
 				}(),
+				BaseRepo: func() (ghrepo.Interface, error) {
+					baseRepo, _ := stubPR("OWNER/REPO:master", "hubot/REPO:feature")
+					return baseRepo, nil
+				},
 				Config: func() (gh.Config, error) {
 					return config.NewBlankConfig(), nil
 				},
@@ -196,6 +212,9 @@ func Test_checkoutRun(t *testing.T) {
 			opts: &CheckoutOptions{
 				SelectorArg: "",
 				Interactive: false,
+				BaseRepo: func() (ghrepo.Interface, error) {
+					return ghrepo.New("OWNER", "REPO"), nil
+				},
 			},
 			remotes: map[string]string{
 				"origin": "OWNER/REPO",
@@ -303,7 +322,7 @@ func Test_checkoutRun(t *testing.T) {
 
 /** LEGACY TESTS **/
 
-func runCommand(rt http.RoundTripper, remotes context.Remotes, branch string, cli string) (*test.CmdOut, error) {
+func runCommand(rt http.RoundTripper, remotes context.Remotes, branch string, cli string, baseRepo ghrepo.Interface) (*test.CmdOut, error) {
 	ios, _, stdout, stderr := iostreams.Test()
 
 	factory := &cmdutil.Factory{
@@ -332,6 +351,9 @@ func runCommand(rt http.RoundTripper, remotes context.Remotes, branch string, cl
 			GhPath:  "some/path/gh",
 			GitPath: "some/path/git",
 		},
+		BaseRepo: func() (ghrepo.Interface, error) {
+			return baseRepo, nil
+		},
 	}
 
 	cmd := NewCmdCheckout(factory, nil)
@@ -369,7 +391,7 @@ func TestPRCheckout_sameRepo(t *testing.T) {
 	cs.Register(`git show-ref --verify -- refs/heads/feature`, 1, "")
 	cs.Register(`git checkout -b feature --track origin/feature`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -391,7 +413,7 @@ func TestPRCheckout_existingBranch(t *testing.T) {
 	cs.Register(`git checkout feature`, 0, "")
 	cs.Register(`git merge --ff-only refs/remotes/origin/feature`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -424,7 +446,7 @@ func TestPRCheckout_differentRepo_remoteExists(t *testing.T) {
 	cs.Register(`git show-ref --verify -- refs/heads/feature`, 1, "")
 	cs.Register(`git checkout -b feature --track robot-fork/feature`, 0, "")
 
-	output, err := runCommand(http, remotes, "master", `123`)
+	output, err := runCommand(http, remotes, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -449,7 +471,7 @@ func TestPRCheckout_differentRepo(t *testing.T) {
 	cs.Register(`git config branch\.feature\.pushRemote origin`, 0, "")
 	cs.Register(`git config branch\.feature\.merge refs/pull/123/head`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -470,7 +492,7 @@ func TestPRCheckout_differentRepo_existingBranch(t *testing.T) {
 	cs.Register(`git config branch\.feature\.merge`, 0, "refs/heads/feature\n")
 	cs.Register(`git checkout feature`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -491,7 +513,7 @@ func TestPRCheckout_detachedHead(t *testing.T) {
 	cs.Register(`git config branch\.feature\.merge`, 0, "refs/heads/feature\n")
 	cs.Register(`git checkout feature`, 0, "")
 
-	output, err := runCommand(http, nil, "", `123`)
+	output, err := runCommand(http, nil, "", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -512,7 +534,7 @@ func TestPRCheckout_differentRepo_currentBranch(t *testing.T) {
 	cs.Register(`git config branch\.feature\.merge`, 0, "refs/heads/feature\n")
 	cs.Register(`git merge --ff-only FETCH_HEAD`, 0, "")
 
-	output, err := runCommand(http, nil, "feature", `123`)
+	output, err := runCommand(http, nil, "feature", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -528,7 +550,7 @@ func TestPRCheckout_differentRepo_invalidBranchName(t *testing.T) {
 	_, cmdTeardown := run.Stub()
 	defer cmdTeardown(t)
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.EqualError(t, err, `invalid branch name: "-foo"`)
 	assert.Equal(t, "", output.Stderr())
 	assert.Equal(t, "", output.Stderr())
@@ -553,7 +575,7 @@ func TestPRCheckout_maintainerCanModify(t *testing.T) {
 	cs.Register(`git config branch\.feature\.pushRemote https://github\.com/hubot/REPO\.git`, 0, "")
 	cs.Register(`git config branch\.feature\.merge refs/heads/feature`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123`)
+	output, err := runCommand(http, nil, "master", `123`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -576,7 +598,7 @@ func TestPRCheckout_recurseSubmodules(t *testing.T) {
 	cs.Register(`git submodule sync --recursive`, 0, "")
 	cs.Register(`git submodule update --init --recursive`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123 --recurse-submodules`)
+	output, err := runCommand(http, nil, "master", `123 --recurse-submodules`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
@@ -597,7 +619,7 @@ func TestPRCheckout_force(t *testing.T) {
 	cs.Register(`git checkout feature`, 0, "")
 	cs.Register(`git reset --hard refs/remotes/origin/feature`, 0, "")
 
-	output, err := runCommand(http, nil, "master", `123 --force`)
+	output, err := runCommand(http, nil, "master", `123 --force`, baseRepo)
 
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
@@ -618,7 +640,7 @@ func TestPRCheckout_detach(t *testing.T) {
 	cs.Register(`git remote get-url origin`, 0, "https://github.com/hubot/REPO.git")
 	cs.Register(`git fetch origin refs/pull/123/head`, 0, "")
 
-	output, err := runCommand(http, nil, "", `123 --detach`)
+	output, err := runCommand(http, nil, "", `123 --detach`, baseRepo)
 	assert.NoError(t, err)
 	assert.Equal(t, "", output.String())
 	assert.Equal(t, "", output.Stderr())
```
</details>

nilvng committed:
issue cli#2329: fix unsufficent properties required for PR checkout

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

BagToad commented on pkg/cmd/pr/checkout/checkout.go:
```diff
@@ -268,3 +295,71 @@ func executeCmds(client *git.Client, credentialPattern git.CredentialPattern, cm
+func promptForPR(prompter shared.Prompter, cs *iostreams.ColorScheme, jobs api.PullRequestAndTotalCount) (*api.PullRequest, error) {
+	candidates := []string{}
+	for _, pr := range jobs.PullRequests {
+		candidates = append(candidates, text.Truncate(120, fmt.Sprintf("%s %s (%s)",
+			shared.PRNumberWithColor(cs, pr),
+			text.RemoveExcessiveWhitespace(pr.Title),
+			cs.Gray(pr.HeadLabel()),
+		)))
```
```suggestion
		candidates = append(candidates, text.Truncate(120, fmt.Sprintf("%s %s [%s]",
			shared.PRNumberWithColor(cs, pr),
			text.RemoveExcessiveWhitespace(pr.Title),
			cs.Gray(pr.HeadLabel()),
		)))
```

💭 I think this should be in square brackets to align with [Primer guidelines on displaying branches](https://primer.style/native/cli/components#branches)

nilvng replied:
gotcha! I will update it per your suggestion 
Initially, I was following the design of the prompt `gh run view`
<img width="696" alt="Screenshot 2024-12-06 at 5 59 55 pm" src="https://github.com/user-attachments/assets/2f954f02-01ee-4148-9278-bbf03d21d4ac">

maybe we can update them for the consistency as well

BagToad replied:
Good catch! 

I opened https://github.com/cli/cli/issues/10038. Assuming the team agrees, it seems like something that would be a good first issue for someone 😁

nilvng replied:
thanks @BagToad! 🙌

BagToad commented on pkg/cmd/pr/checkout/checkout.go:
```diff
@@ -42,17 +48,32 @@ func NewCmdCheckout(f *cmdutil.Factory, runF func(*CheckoutOptions) error) *cobr
-		Use:   "checkout {<number> | <url> | <branch>}",
+		Use:   "checkout [<number> | <url> | <branch>]",
 		Short: "Check out a pull request in git",
-		Args:  cmdutil.ExactArgs(1, "argument required"),
```
💭 I think we will want to use `cobra.MaximumNArgs` ([examples](https://github.com/search?q=repo%3Acli%2Fcli+cobra.MaximumNArgs%28&type=code)) here:

```Go
Args:  cobra.MaximumNArgs(1),
```

BagToad committed:
Merge branch 'trunk' into nil/fix-2329

BagToad committed:
issue cli#2329: Fix linter

nilvng committed:
issue #2329: fix unit tests

nilvng committed:
issue #2329: update the prompt's format

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

nilvng committed:
issue #2329: simplify the UI of the prompt

This commit reverts the previous color changes in the prompt UI. While color highlighting could potentially improve the visual appealing of the prompt
using the existing color library (mguz/ansi) with the prompt library (AlecAivazis/survey) caused unintended side effects.
It reset the bold text style for the selected option. We decide to that
bold text style would have a higher priority than the color text ,for
three reasons:
1. To maintain consistency with other prompts in the UI and prioritize accessibility
2. While color can enhance the user experience, according to Primer Design Guidelines, color should not be relied upon to convey essential information.
3. visual indicator of the selected option, especially crucial when dealing with long PR titles or branch names.

As a future improvement, we may consider a separate issue or PR to address the color library issue and explore controlled color usage in prompts. This could potentially allow for more nuanced visual differentiation while avoiding unintended style resets.

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

nilvng committed:
issue #2329: fix prompt header getting duplicated

following the approach in gh run view seems to fix this issue

nilvng committed:
issue #2329: include PR status in the prompt's options

nilvng commented:
hey @BagToad cc @williammartin,  The core functionality of the PR is now stable 🚀 
(huge shoutout to @BagToad for taking time soundboard it with me) 🙇 
 let me know if you have any questions or concerns, e.g. I'm particularly interested in your thoughts on the baseRepo variable in the legacy unit tests.
I'm going to think through how we might refactor the `ListPullRequest()` and unit tests per the scope we have defined here

nilvng committed:
Merge branch 'trunk' into nil/fix-2329

BagToad commented:
👋 Hey @nilvng - thanks for your continued work and collaboration on this PR ✨ Thank you for updating the color as we discussed and working on the unit tests.

Also wanted to mention that adding the state to the prompt seems like a good idea to me 😁 

I'm going to try and address a few things that should hopefully unblock some further development of this feature pending a larger team review, but let me know if I missed anything that you need from us ❤️ 

---

>  I'm particularly interested in your thoughts on the baseRepo variable in the legacy unit tests. 

I'm not sure about this one 🤔 I think I'd like to hear the rest of the team's thoughts on this and legacy checkout tests once they have some time.

> I'm going to think through how we might refactor the ListPullRequest() and unit tests per the scope we have defined here

Sounds good! I thought a lot about this, and I'm not sure if I know the best solution - we might need to iterate on it. I think we may want to start with following how `Finder` is implemented and iterate from there; create a new `PRLister` interface, implementation, and mock, to inject into our `Opts`.

 If I could soundboard my thoughts on what we are testing, I think our testing goals are...

- Test that `checkoutRun` is properly using a to-be-created `PRLister` interface correctly by including `MaintainerCanModify` and `HeadRepository` in the query 🤔 
- Our mock `PRLister` should have a way to check that these values were requested, then we should assert that in the tests.

It leaves a bit to be desired, but let's start there and perhaps the rest of the team will have more feedback as we go. I hope this unblocks you to work on that refactor when desired 😁 

---

> let me know if you have any questions or concerns

I do have some things I've been thinking about... 🤔 

### 💭 We should consider a "zero state" message

**Current**

```
❯ gh pr checkout
could not prompt: please provide options to select from
```

**Desired (example from `run delete`)**

```
❯ gh run delete
found no runs to delete
```

> [!NOTE]
> More inconsistency here as the "zero state" for `run view` is the same 😭 🙃 
> 
> ```
> ❯ ghrun view
> could not prompt: please provide options to select from
> ```
> 
> Nonetheless, I think if we're doing this right, I think we should have a message for when there are no PRs found.

### 💭 Listing only 10 PRs feels like it might not be enough

> [!NOTE]
> I don't think any of this blocks this PR. In fact, I'm not convinced this feedback needs to be addressed in this PR. Just noting my concerns for feedback.

I don't know what the right solution is for this, but I wanted to express how this might not be the optimal prompting experience for some folks (like our favorite repo, the `cli/cli` repo 🤪 ) that have more than 10 PRs.

Just brainstorming some things here:

**Allow a `--limit` flag to control how many PRs are listed in the prompt**
 
Problems: this isn't a precedent that I could find; I cannot find other commands allowing the number of items prompted to be adjusted via a flag like this. 

Something that concerns me is the UX of having a flag that _only does stuff for the prompting_. I think this means that if that's something the team is okay with, we'd want to be very clear with the flag naming.

**Don't list draft PRs in the prompt by default, maybe permit listing drafts with a flag**

Problems: Same problems as above regarding flags adjusting the content of _only_ the prompt list. 

Personal opinion - generally, the PRs I want to checkout are going to be open with some exceptions that I would be happy to pass a flag for. Maybe this behavior could be seen as cumbersome for some that work with drafts often and would dislike that default behavior.

Edit: It's worth mentioning that the reason I bring this up is because _excluding drafts would help me squeeze more value out of those 10 PRs that I get to see_

**Create an environment variable to adjust the number of prompt items for commands that support it including commands like `run view`**

Problems: Perhaps discoverability of an environment variable leaves some to be desired, but I don't think that's a reason to avoid this option. Worth mentioning is that we do support env vars like `GH_MDWIDTH`, which are adjacently related in that we have an env for adjusting formatting related concerns.

nilvng committed:
Merge branch 'trunk' into nil/fix-2329

nilvng committed:
issue #2329: handle empty list of PR

nilvng committed:
fix: tab format may break checkout prompts

nilvng committed:
issue #2329: create shared PRLister

nilvng committed:
feat: integrate new PRLister into pr list command

nilvng committed:
chore: remove unused file

williammartin commented:
@nilvng, @BagToad is on vacation now. I'm sort of loosely following along, let me know if/when you think this is in a position for me to review.

nilvng commented:
@williammartin Thanks for checking in! I'd really appreciate your feedback on the PR. I think it's in a good state for review, though I'm still working through a couple of open questions.

For context, here's what I've been working on recently:
1. Improved Testability of PR required fields for checkout: Refactored `ListPullRequest()` into the `PRLister` interface, similar to the existing `Finder` approach. similar to the existing Finder approach.
2. Handle the "zero-state" message - where there are no PRs available - we will reuse the message in `shared.ListNoResults()` e.g. `Sprintf("no open %ss in %s", itemName, repoName)`

There are also things still are quite up in the air:
1. Move the GraphQL requests to the `api/` module - as it is one mentioned by @BagToad [ref](https://github.com/cli/cli/pull/9868#discussion_r1864210477) Given now that we already have had the PRLister that is shared within the `pr/` module,  I'm weighing the benefits of improved organization (moving to api/) against keeping things close to where they're used in pr/. What unit tests, and mocks I would need to move along as well

2. The hard limit of 10 PRs leaves to be desired - I'm still pondering over @BagToad suggestions for a quick win that addresses most use cases without bloating the primary feature of `gh pr checkout`  I would tend to believe that we can iterate on this topic later, but at the same time, I agree with @BagToad  that 10 is not optimal solution - While it's sufficient for a smaller startup like mine, I'm concerned it might defeat the usefulness of the feature in larger repositories. super keen for your perspectives on this matter!

nilvng committed:
Merge branch 'trunk' into nil/fix-2329

BagToad commented:
👋 Hey @nilvng, thank you for your patience on this. We're getting caught up after the holidays 😁 

> * Improved Testability of PR required fields for checkout: Refactored `ListPullRequest()` into the `PRLister` interface, similar to the existing `Finder` approach. similar to the existing Finder approach.
> * Handle the "zero-state" message - where there are no PRs available - we will reuse the message in `shared.ListNoResults()` e.g. `Sprintf("no open %ss in %s", itemName, repoName)`

Great work! ✨ 

I discussed this PR and your questions with @williammartin:

> 1. Move the GraphQL requests to the `api/` module - as it is one mentioned by @BagToad [ref](https://github.com/cli/cli/pull/9868#discussion_r1864210477) Given now that we already have had the PRLister that is shared within the `pr/` module,  I'm weighing the benefits of improved organization (moving to api/) against keeping things close to where they're used in pr/. What unit tests, and mocks I would need to move along as well

Neither of us see much value in moving it to an `api` package right now. Let's keep it beside the `PRLister` in the `pr` package👍 

> 2\. The hard limit of 10 PRs leaves to be desired - I'm still pondering over @BagToad suggestions for a quick win that addresses most use cases without bloating the primary feature of `gh pr checkout`  I would tend to believe that we can iterate on this topic later, but at the same time, I agree with @BagToad  that 10 is not optimal solution - While it's sufficient for a smaller startup like mine, I'm concerned it might defeat the usefulness of the feature in larger repositories. super keen for your perspectives on this matter!

@williammartin and I both agreed that, while there may be improvement opportunities here, it shouldn't block this PR from being merged. So, we'll "ship to learn" on this one 😁 I'm excited to add this feature to my workflow for a few months to really see if the 10 PR limit is even that much of an issue 💭

With all that in mind, I'll work on giving this another review ❤️

BagToad commented on pkg/iostreams/color.go:
```diff
@@ -252,5 +252,5 @@ func (c *ColorScheme) HexToRGB(hex string, x string) string {
 	g, _ := strconv.ParseInt(hex[2:4], 16, 64)
 	b, _ := strconv.ParseInt(hex[4:6], 16, 64)
-	return fmt.Sprintf("\033[38;2;%d;%d;%dm%s\033[0m", r, g, b, x)
+	return fmt.Sprintf("\033[38;2;%d;%d;%dm%s\033[39m", r, g, b, x)
```
@nilvng Since we reverted the color options in this PR, can we revert this change as well? 

Same with L21 & L65 in pkg/iostreams/color_test.go

nilvng replied:
yes :)

BagToad commented on pkg/cmd/pr/shared/finder.go:
```diff
@@ -568,7 +568,7 @@ func (m *mockFinder) ExpectFields(fields []string) {
 }
 
-func isEqualSet(a, b []string) bool {
+func IsEqualSet(a, b []string) bool {
```
💭 Can `IsEqualSet` be un-exported because it is only used in the `shared` package?

nilvng replied:
fixed! thanks :)

BagToad commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -19,6 +19,10 @@ import (
 
 type InputType int
 
+type Prompter interface {
+	Select(string, string, []string) (int, error)
+}
+
```
💭 I'm contemplating the differences between this and the `Prompt` interface defined in pr/shared/survey.go: https://github.com/cli/cli/blob/7fcb4453ed86e949cce50814b268ff86a5052bbf/pkg/cmd/pr/shared/survey.go#L36-L42

❓ Do you think we can just use that existing interface instead of redefining an interface here? Either way, I'm thinking the interface definition should be in pr/shared/survey.go 🤔 

What do you think?

nilvng replied:
Good point!  I have updated the code to be re-using `Prompt` type from `pr/shared/survey.go` 🙌

BagToad commented on pkg/cmd/pr/checkout/checkout_test.go:
```diff
@@ -183,12 +212,87 @@ func Test_checkoutRun(t *testing.T) {
+						PullRequests: []api.PullRequest{
+							*pr1, *pr2, *pr3,
+						}, SearchCapped: false}, nil)
+					lister.ExpectFields([]string{"number", "title", "state", "isDraft", "headRefName", "headRepository", "headRepositoryOwner", "isCrossRepository", "maintainerCanModify"})
```
Love to see that we're checking for these! ✨

nilvng committed:
Revert color changes

This reverts commit 36eaf14857dc8110c47b3d91b823589771aae305.

nilvng committed:
revert isEqualSet to private

nilvng committed:
remove duplicated Prompter type

BagToad committed:
Merge branch 'trunk' into nil/fix-2329

BagToad reviewed (approved):
👋 Hey @nilvng - thank you for all the effort and collaboration in this PR ✨ ❤️ Apologies it took so long! I appreciate you taking the leap into implementing an old issue like this and tackling so many unknowns along the way.

I've done another thorough pass, and I think everything looks good and functions well! 😁 🚀 

Thank you again for all the work, patience, ideas, and attention to detail ❤️

nilvng commented:
Hi @BagToad 👋 
It's my pleasure to collaborate with you on this little project! Thank you for reviewing my pull request; your feedback was useful and helped me better understand the codebase 🤩 I'm glad to hear that everything looks good, and I look forward to contributing more to the repository in the future.
